src/uct/tcp/tcp.h

```diff
@@ -400,6 +400,7 @@ typedef struct uct_tcp_iface {
         struct sockaddr_storage   ifaddr;            /* Network address */
         struct sockaddr_storage   netmask;           /* Network address mask */
         size_t                    sockaddr_len;      /* Network address length */
+        int                       ep_bind_src_addr;  /* Bind EP's FD to ifaddr */
```

Comment:
> * I was able to reproduce the issue with `rp_filter=2` without binding
> * `UCX_TCP_EP_BIND_SRC_ADDR` is enabled by default and the issue seems to be fixed on my setup, do you see the case when it may have unexpected effect?
> 
> AFAIU `rp_filter` is a system setting which cannot be modified from user space UCX environment, I guess the only option that we can do with it - check and print debug/diag/warning message if `rp_filter=1`

I don't see any other case of unexpected effect currently . It would be better if the message can provide more info about the risk of changing rp_filter setting, may be just a link is enough: https://www.[kernel.org/doc/Documentation/networking/ip-sysctl.txt](https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt) .

> 
> IMO fds consumption is critical on scale and using exact iface address is exactly what UCT layer should do, so binding approach looks good to me. 

Agree.

Oh, BTW,  you wrote: fixes https://github.com/openucx/ucx/issues/9763 miltirail case with multiple network interfaces in the same subnet.    But this issue can also occured when those interfaces are in different subnets.

Reply:
> But this issue can also occured when those interfaces (of one node) are in different subnets.

I'm not completely understand how it might happen in different subnet if interfaces are reachable only pairwise. But if you can confirm that this PR helps for you case then I just remove "same subnet" from the commit message.

Thank you for your cooperation!

Reply:
Also I dont think that hardcode a link to web documentation which can be changed any time is the best practice, maybe better just mention "see RFC3704 for details"? @yosefe WDYT?

Reply:
> > But this issue can also occured when those interfaces (of one node) are in different subnets.
> 
> I'm not completely understand how it might happen in different subnet if interfaces are reachable only pairwise. But if you can confirm that this PR helps for you case then I just remove "same subnet" from the commit message.
> 

I confirmed that this issue also occured when those interfaces (of one node) are in different subnets with ucx release 1.15. But with release 1.16, I can not reproduce this issue with interfaces (of one node) are in different subnets any more. It seems some changes during1.15 to 1.16 omits this issue by re-arranging the ifaces mapping for lane[0] and lane[1] which makes ucp_wireup_connect_lane() uses right remote address to connect to at the first place, thus, no subnet cross over.

Below is debug log of 1.16 vs 1.15 from which you can see that 1.15 chose ens8f0 for lane[0], while 1.16 chose ens7f0 for lane[0]. So I think if this is an intentinal change to solve subnet cross over connection problem, then "same subnet" can be removed from the commit message.

1.16

```
[1713927611.038973] [promote:39009:0]      ucp_worker.c:1887 UCX  INFO    perftest inter-node cfg#0 am(tcp/enp7s0 tcp/enp8s0)
[1713927611.038977] [promote:39009:0]          wireup.c:1236 UCX  DEBUG   ep 0x7fadd7f42000: am_lane 0 wireup_msg_lane 1 cm_lane <none> keepalive_lane <none> reachable_mds 0x1
[1713927611.038981] [promote:39009:0]          wireup.c:1246 UCX  DEBUG   ep 0x7fadd7f42000: lane[0]:  0:tcp/enp7s0.0 md[0]            -> addr[0].md[0]/tcp/sysdev[0] seg 4294967295 rma_bw#1 am am_bw#0
[1713927611.038984] [promote:39009:0]          wireup.c:1246 UCX  DEBUG   ep 0x7fadd7f42000: lane[1]:  1:tcp/enp8s0.0 md[0]            -> addr[1].md[0]/tcp/sysdev[1] seg 4294967295 rma_bw#0 am_bw#1 wireup
[1713927611.038985] [promote:39009:0]          wireup.c:1250 UCX  DEBUG   ep 0x7fadd7f42000: err mode 0, flags 0x0
[1713927611.038987] [promote:39009:0]          wireup.c:1166 UCX  TRACE   ep 0x7fadd7f42000: connect lane[0]
[1713927611.038989] [promote:39009:0]          wireup.c:1054 UCX  TRACE   ep 0x7fadd7f42000: connect uct_ep[0] to addr 0x16f9410
[1713927611.039004] [promote:39009:0]          tcp_ep.c:259  UCX  DEBUG   tcp_ep 0x172b490: created on iface 0x16b1d20, fd -1
[1713927611.039014] [promote:39009:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x172b490: CLOSED -> CONNECTING for the [192.168.100.2:36757]<->[192.168.100.1:36877]:0 connection [-:-]
[1713927611.039021] [promote:39009:0]      conn_match.c:269  UCX  TRACE   match_ctx 0x16b2228: address 192.168.100.1:36877 conn_sn 0 not found in unexpected
[1713927611.039052] [promote:39009:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x172b490: CONNECTING -> CONNECTING for the [192.168.100.2:36757]<->[192.168.100.1:36877]:0 connection [-:-]
[1713927611.039259] [promote:39009:0]            sock.c:347  UCX  DEBUG   connect(fd=14, src_addr=192.168.100.2:58298 dest_addr=192.168.100.1:36877): Success
[1713927611.039284] [promote:39009:0]          tcp_cm.c:144  UCX  TRACE   tcp_ep 0x172b490: UNKNOWN (1) [192.168.100.1:36877]:0
[1713927611.039288] [promote:39009:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x172b490: CONNECTING -> WAITING_ACK for the [192.168.100.2:36757]<->[192.168.100.1:36877]:0 connection [-:-]
[1713927611.039290] [promote:39009:0]          tcp_ep.c:911  UCX  TRACE   tcp_ep 0x172b490: set events to r-
[1713927611.039311] [promote:39009:0]          tcp_ep.c:281  UCX  TRACE   tcp_ep 0x172b490: ctx caps changed [-:-] -> [Tx:-]
[1713927611.039320] [promote:39009:0]      conn_match.c:197  UCX  TRACE   match_ctx 0x16b2228: conn_match 0x172b510 added as expected address 192.168.100.1:36877 conn_sn 0
[1713927611.039323] [promote:39009:0]          wireup.c:1092 UCX  TRACE   ep 0x7fadd7f42000: assign uct_ep[0]=0x172b490
[1713927611.039326] [promote:39009:0]      ucp_worker.c:575  UCX  TRACE   activate iface 0x16b1d20 (tcp/enp7s0) acount=3 aifaces=2
[1713927611.039327] [promote:39009:0]          wireup.c:1166 UCX  TRACE   ep 0x7fadd7f42000: connect lane[1]
[1713927611.039329] [promote:39009:0]          wireup.c:1054 UCX  TRACE   ep 0x7fadd7f42000: connect uct_ep[1] to addr 0x16f9618
[1713927611.039332] [promote:39009:0]          tcp_ep.c:259  UCX  DEBUG   tcp_ep 0x172b6b0: created on iface 0x16b24a0, fd -1
[1713927611.039336] [promote:39009:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x172b6b0: CLOSED -> CONNECTING for the [192.168.200.2:42523]<->[192.168.200.1:55285]:0 connection [-:-]
[1713927611.039339] [promote:39009:0]      conn_match.c:269  UCX  TRACE   match_ctx 0x16b29a8: address 192.168.200.1:55285 conn_sn 0 not found in unexpected
[1713927611.039354] [promote:39009:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x172b6b0: CONNECTING -> CONNECTING for the [192.168.200.2:42523]<->[192.168.200.1:55285]:0 connection [-:-]
[1713927611.039484] [promote:39009:0]            sock.c:347  UCX  DEBUG   connect(fd=15, src_addr=192.168.200.2:47052 dest_addr=192.168.200.1:55285): Success
[1713927611.039501] [promote:39009:0]          tcp_cm.c:144  UCX  TRACE   tcp_ep 0x172b6b0: UNKNOWN (1) [192.168.200.1:55285]:0
[1713927611.039505] [promote:39009:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x172b6b0: CONNECTING -> WAITING_ACK for the [192.168.200.2:42523]<->[192.168.200.1:55285]:0 connection [-:-]
[1713927611.039506] [promote:39009:0]          tcp_ep.c:911  UCX  TRACE   tcp_ep 0x172b6b0: set events to r-
[1713927611.039511] [promote:39009:0]          tcp_ep.c:281  UCX  TRACE   tcp_ep 0x172b6b0: ctx caps changed [-:-] -> [Tx:-]
[1713927611.039517] [promote:39009:0]      conn_match.c:197  UCX  TRACE   match_ctx 0x16b29a8: conn_match 0x172b730 added as expected address 192.168.200.1:55285 conn_sn 0
[1713927611.039546] [promote:39009:0]          wireup.c:1092 UCX  TRACE   ep 0x7fadd7f42000: assign uct_ep[1]=0x172b6b0
[1713927611.039548] [promote:39009:0]      ucp_worker.c:575  UCX  TRACE   activate iface 0x16b24a0 (tcp/enp8s0) acount=1 aifaces=2
```



1.15

```

[1713929603.574968] [promote:41075:0]          wireup.c:1204 UCX  DEBUG   ep 0x7f8fcb15a000: am_lane 0 wireup_msg_lane 1 cm_lane <none> keepalive_lane <none> reachable_mds 0x1
[1713929603.574972] [promote:41075:0]          wireup.c:1214 UCX  DEBUG   ep 0x7f8fcb15a000: lane[0]:  0:tcp/enp8s0.0 md[0]            -> addr[0].md[0]/tcp/sysdev[255] rma_bw#1 am am_bw#0
[1713929603.574975] [promote:41075:0]          wireup.c:1214 UCX  DEBUG   ep 0x7f8fcb15a000: lane[1]:  1:tcp/enp7s0.0 md[0]            -> addr[1].md[0]/tcp/sysdev[255] rma_bw#0 am_bw#1 wireup
[1713929603.574976] [promote:41075:0]          wireup.c:1218 UCX  DEBUG   ep 0x7f8fcb15a000: err mode 0, flags 0x0
[1713929603.574978] [promote:41075:0]          wireup.c:1134 UCX  TRACE   ep 0x7f8fcb15a000: connect lane[0]
[1713929603.574987] [promote:41075:0]          wireup.c:1028 UCX  TRACE   ep 0x7f8fcb15a000: connect uct_ep[0] to addr 0x26ac090
[1713929603.575014] [promote:41075:0]          tcp_ep.c:259  UCX  DEBUG   tcp_ep 0x26a9f60: created on iface 0x2646b30, fd -1
[1713929603.575027] [promote:41075:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x26a9f60: CLOSED -> CONNECTING for the [192.168.200.2:39533]<->[192.168.100.1:52461]:0 connection [-:-]
[1713929603.575033] [promote:41075:0]      conn_match.c:269  UCX  TRACE   match_ctx 0x2647038: address 192.168.100.1:52461 conn_sn 0 not found in unexpected
[1713929603.575059] [promote:41075:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x26a9f60: CONNECTING -> CONNECTING for the [192.168.200.2:39533]<->[192.168.100.1:52461]:0 connection [-:-]
[1713929603.575529] [promote:41075:0]            sock.c:336  UCX  DEBUG   connect(fd=14, src_addr=192.168.100.2:53056 dest_addr=192.168.100.1:52461): Success
[1713929603.575553] [promote:41075:0]          tcp_cm.c:144  UCX  TRACE   tcp_ep 0x26a9f60: UNKNOWN (1) [192.168.100.1:52461]:0
[1713929603.575559] [promote:41075:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x26a9f60: CONNECTING -> WAITING_ACK for the [192.168.200.2:39533]<->[192.168.100.1:52461]:0 connection [-:-]
[1713929603.575565] [promote:41075:0]          tcp_ep.c:934  UCX  TRACE   tcp_ep 0x26a9f60: set events to r-
[1713929603.575577] [promote:41075:0]          tcp_ep.c:281  UCX  TRACE   tcp_ep 0x26a9f60: ctx caps changed [-:-] -> [Tx:-]
[1713929603.575587] [promote:41075:0]      conn_match.c:197  UCX  TRACE   match_ctx 0x2647038: conn_match 0x26a9fe0 added as expected address 192.168.100.1:52461 conn_sn 0
[1713929603.575592] [promote:41075:0]          wireup.c:1066 UCX  TRACE   ep 0x7f8fcb15a000: assign uct_ep[0]=0x26a9f60
[1713929603.575596] [promote:41075:0]      ucp_worker.c:566  UCX  TRACE   activate iface 0x2646b30 acount=0 aifaces=0
[1713929603.576993] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp8s0: pci bandwidth undetected, using maximal value
[1713929603.578076] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp8s0: pci bandwidth undetected, using maximal value
[1713929603.579109] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp8s0: pci bandwidth undetected, using maximal value
[1713929603.580124] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp8s0: pci bandwidth undetected, using maximal value
[1713929603.581075] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp8s0: pci bandwidth undetected, using maximal value
[1713929603.582187] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp8s0: pci bandwidth undetected, using maximal value
[1713929603.583189] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp8s0: pci bandwidth undetected, using maximal value
[1713929603.584296] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp8s0: pci bandwidth undetected, using maximal value
[1713929603.585359] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp8s0: pci bandwidth undetected, using maximal value
[1713929603.586355] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp8s0: pci bandwidth undetected, using maximal value
[1713929603.586417] [promote:41075:0]          wireup.c:1134 UCX  TRACE   ep 0x7f8fcb15a000: connect lane[1]
[1713929603.586420] [promote:41075:0]          wireup.c:1028 UCX  TRACE   ep 0x7f8fcb15a000: connect uct_ep[1] to addr 0x26ac2a0
[1713929603.586424] [promote:41075:0]          tcp_ep.c:259  UCX  DEBUG   tcp_ep 0x2648870: created on iface 0x2663480, fd -1
[1713929603.586430] [promote:41075:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x2648870: CLOSED -> CONNECTING for the [192.168.100.2:42833]<->[192.168.200.1:34651]:0 connection [-:-]
[1713929603.586434] [promote:41075:0]      conn_match.c:269  UCX  TRACE   match_ctx 0x2663988: address 192.168.200.1:34651 conn_sn 0 not found in unexpected
[1713929603.586446] [promote:41075:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x2648870: CONNECTING -> CONNECTING for the [192.168.100.2:42833]<->[192.168.200.1:34651]:0 connection [-:-]
[1713929603.586549] [promote:41075:0]            sock.c:336  UCX  DEBUG   connect(fd=15, src_addr=192.168.200.2:55370 dest_addr=192.168.200.1:34651): Success
[1713929603.586568] [promote:41075:0]          tcp_cm.c:144  UCX  TRACE   tcp_ep 0x2648870: UNKNOWN (1) [192.168.200.1:34651]:0
[1713929603.586572] [promote:41075:0]          tcp_cm.c:104  UCX  DEBUG   tcp_ep 0x2648870: CONNECTING -> WAITING_ACK for the [192.168.100.2:42833]<->[192.168.200.1:34651]:0 connection [-:-]
[1713929603.586573] [promote:41075:0]          tcp_ep.c:934  UCX  TRACE   tcp_ep 0x2648870: set events to r-
[1713929603.586580] [promote:41075:0]          tcp_ep.c:281  UCX  TRACE   tcp_ep 0x2648870: ctx caps changed [-:-] -> [Tx:-]
[1713929603.586587] [promote:41075:0]      conn_match.c:197  UCX  TRACE   match_ctx 0x2663988: conn_match 0x26488f0 added as expected address 192.168.200.1:34651 conn_sn 0
[1713929603.586590] [promote:41075:0]          wireup.c:1066 UCX  TRACE   ep 0x7f8fcb15a000: assign uct_ep[1]=0x2648870
[1713929603.586592] [promote:41075:0]      ucp_worker.c:566  UCX  TRACE   activate iface 0x2663480 acount=0 aifaces=1
[1713929603.587756] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp7s0: pci bandwidth undetected, using maximal value
[1713929603.588799] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp7s0: pci bandwidth undetected, using maximal value
[1713929603.589752] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp7s0: pci bandwidth undetected, using maximal value
[1713929603.590701] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp7s0: pci bandwidth undetected, using maximal value
[1713929603.591653] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp7s0: pci bandwidth undetected, using maximal value
[1713929603.592609] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp7s0: pci bandwidth undetected, using maximal value
[1713929603.593560] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp7s0: pci bandwidth undetected, using maximal value
[1713929603.594506] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp7s0: pci bandwidth undetected, using maximal value
[1713929603.595460] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp7s0: pci bandwidth undetected, using maximal value
[1713929603.596412] [promote:41075:0]            topo.c:742  UCX  DEBUG   enp7s0: pci bandwidth undetected, using maximal value

```

Reply:
> Also I dont think that hardcode a link to web documentation which can be changed any time is the best practice, maybe better just mention "see RFC3704 for details"? @yosefe WDYT?

IMO kernel.org link is ok, we can also mention RFC number